DM-46581: Speed up TimeConverter

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Change Timespan repr for jd times to maintain precision

Include the jd1 and jd2 to ensure that we can recreate
the time accurately enough from the repr() form.

timj committed:
Speed up nsec_to_astropy by avoiding TimeDelta

Constructing the TimeDelta is slow. Calculating the JD components
and explicitly adding them to the epoch JD values is much faster
than constructing a TimeDelta and adding that to a Time.

timj committed:
Use ._time.jd1 for slight speed increase in astropy_to_nsec

timj commented:
This change almost halves the time of `nsec_to_astropy` (and makes `astropy_to_nsec` 4% faster).

andy-slac committed:
Implement new astropy time format which skips expensive checks.

This format is only ever use by TimeConverter so we can trust that
the values passed are always OK and skip usual checks.

timj committed:
Create Time objects with precision=6

We would like to get microseconds in our ISO output form.

timj committed:
Add explicit dependency on numpy

We only use it in a couple of places explicitly but it is
used.

timj commented on python/lsst/daf/butler/_timespan.py:
```diff
@@ -335,14 +335,23 @@ def __str__(self) -> str:
+        # For JD times we want to use jd1 and jd2 to maintain precision.
+        if isinstance(t, astropy.time.Time):
+            if t.format == "jd":
+                return f"astropy.time.Time({t.jd1}, {t.jd2}, scale='{t.scale}', format='{t.format}')"
```
This doesn't trigger with the new code because for Timespan we always convert the Time to nsec internally and then convert back to Time on request and so return a format=unix_tai_fast variant for repr. If we ever decide to simply keep the Timespan we were given then this would trigger again.

timj commented on python/lsst/daf/butler/time_utils.py:
```diff
@@ -136,10 +177,10 @@ def nsec_to_astropy(self, time_nsec: int) -> astropy.time.Time:
-        return value
+        jd1, jd2 = divmod(time_nsec, 1_000_000_000)
+        return astropy.time.Time(
+            float(jd1), jd2 / 1_000_000_000, format="unix_tai_fast", scale="tai", precision=6
```
@andy-slac  We have to pass in seconds here for jd1 rather than days (and so require the divmod call again) because we are subclassing `TimeUnixTai`?

andy-slac replied:
Right, I decided that we want to use TimeUnixTai instead of TimeJD format as a base because this is exactly what we do, and TimeUnix takes seconds. It would still be possible to pass days here, but I was worried it could cause confusion somewhere else.

andy-slac commented on python/lsst/daf/butler/_timespan.py:
```diff
@@ -335,14 +335,23 @@ def __str__(self) -> str:
+            if t.format == "jd":
+                return f"astropy.time.Time({t.jd1}, {t.jd2}, scale='{t.scale}', format='{t.format}')"
+            else:
+                return f"astropy.time.Time('{t}', scale='{t.scale}', format='{t.format}')"
```
Hmm, this will now render as:
```
astropy.time.Time('1727969332.139023', scale='tai', format='unix_tai_fast')
```
with value appearing as string. Also I am a bit worried that we expose `unix_tai_fast` which is supposed to be internal.

timj replied:
And yet the tests pass using it. It worked "correctly" before because we were using an isot string format by default by defining `self.epoch` as that in `TimeConverter`.

I was also worried about `unix_tai_fast` leaking out but I tried adding a `.replicate(format="jd")` to the `nsec_to_astropy` method and whilst that worked, it did slow things down by factor 2 from 7 to 14 microseconds. Not sure what to do really. We could force to JD or ISOT format in this repr by calling `.replicate()`. The format will still leak out for anyone who decides to print `ts.begin` anywhere.

timj replied:
Oh, and that quoting isn't new. It was always like that.

andy-slac replied:
Maybe instead of replicate() we can do `time.format = "jd"` inside nsec_to_astropy, this makes it about 25% slower.

timj replied:
Ok. I will push that as a compromise.

timj replied:
Slowed down from 8us to 10us.

timj committed:
Return the nsec to astropy time in format JD

We have decided that returning it in tai unix fast format is
going to be confusing since these times are visible to people
via Timespan. Therefore switch the format to "jd" and accept
a small overhead going from 8us to 10us in construction.
